nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py

```diff
@@ -0,0 +1,222 @@
+        t_map_to_base_link.transform.translation.y = self.initial_pose.position.y
+        t_map_to_base_link.transform.translation.z = self.initial_pose.position.z
+        t_map_to_base_link.transform.rotation = self.initial_pose.orientation
+        mat_map_to_base_link = transformStampedToMatrix(t_map_to_base_link)
+        mat_odom_to_base_link = transformStampedToMatrix(self.t_odom_to_base_link)
+        mat_base_link_to_odom = tf_transformations.inverse_matrix(mat_odom_to_base_link)
+        mat_map_to_odom = \
+            tf_transformations.concatenate_matrices(mat_map_to_base_link, mat_base_link_to_odom)
+        self.t_map_to_odom.transform = matrixToTransform(mat_map_to_odom)
+
+    def timerCallback(self):
+        # If no data, just republish existing transforms without change
+        one_sec = Duration(seconds=1)
+        if self.curr_cmd_vel is None or self.get_clock().now() - self.curr_cmd_vel_time > one_sec:
+            self.publishTransforms(self.t_map_to_odom, self.t_odom_to_base_link)
+            self.curr_cmd_vel = None
+            return
+
+        # Update odom->base_link from cmd_vel
+        dx = self.curr_cmd_vel.linear.x * self.update_dur
+        dy = self.curr_cmd_vel.linear.y * self.update_dur
+        dth = self.curr_cmd_vel.angular.z * self.update_dur
+        q = [self.t_odom_to_base_link.transform.rotation.x,
+             self.t_odom_to_base_link.transform.rotation.y,
+             self.t_odom_to_base_link.transform.rotation.z,
+             self.t_odom_to_base_link.transform.rotation.w]
+        _, _, yaw = tf_transformations.euler_from_quaternion(q)
+        self.t_odom_to_base_link.transform.translation.x += dx * math.cos(yaw) - dy * math.sin(yaw)
+        self.t_odom_to_base_link.transform.translation.y += dx * math.sin(yaw) + dy * math.cos(yaw)
+        self.t_odom_to_base_link.transform.rotation = \
+            addYawToQuat(self.t_odom_to_base_link.transform.rotation, dth)
+
+        self.publishTransforms(self.t_map_to_odom, self.t_odom_to_base_link)
+        self.publishOdometry(self.t_odom_to_base_link)
+        self.publishLaserScan()
+
+    def publishLaserScan(self):
+        # Publish a bogus laser scan for collision monitor
+        scan = LaserScan()
+        # scan.header.stamp = (self.get_clock().now()).to_msg()
+        scan.header.frame_id = self.scan_frame_id
+        scan.angle_min = -math.pi
+        scan.angle_max = math.pi
+        scan.angle_increment = 0.005817705996  # 0.333 degrees
+        scan.time_increment = 0.0
+        scan.scan_time = 0.1
+        scan.range_min = 0.1
+        scan.range_max = 100.0
+        num_samples = int((scan.angle_max - scan.angle_min) / scan.angle_increment)
+        scan.ranges = [scan.range_max - 0.01] * num_samples
+        self.scan_pub.publish(scan)
+
+    def publishTransforms(self, map_to_odom, odom_to_base_link):
+        map_to_odom.header.stamp = \
+            (self.get_clock().now() + Duration(seconds=self.update_dur)).to_msg()
```

Comment:
This is how all the global localizers work. We need the transformation from map->odom to slightly "lead" the system so that anyone that wants to transform map->XYZ frame gets that data available between global localization updates. Since that is a static transformation between updates, the time isn't really all that important (unlike the odom->base_link). If you look at Cartographer, AMCL, SLAM Toolbox, etc they all do this from the very beginning of ROS 1 as a known design pattern to prevent floods of those "extrapolation into the future" errors

Reply:
Thanks for the explanation.

We have just killed off an controller_server control loop issue that is probably related to this. 
i.e  "Control loop missed its desired rate of %.4f Hz. Current loop rate is %.4f Hz."  https://github.com/ros-navigation/navigation2/blob/3f311124e356dbfa33f35e23caa5474a57f99ab7/nav2_controller/src/controller_server.cpp#L498

Despite configuring for rtprio and we were still getting this warning 

I think MPPI is effectively hanging around waiting for odom updates.  We were running MPPI at 50Hz and the two ros_localisation ekfs at 30Hz and nav_sat_transform at 30Hz.  So while it works I have a feeling there is effectively a live lock underneath it all.  

Do you, in your robots change the robot_localization transform_time_offset from the default of 0.0?  https://github.com/cra-ros-pkg/robot_localization/blob/f6b28e0c5bcf98d161cb65b9d28d00ced77fe3ab/params/ekf.yaml#L20

Bumping the ekfs to 100Hz and leaving nav_sat_transform at 30Hz makes nav2_controller/MPPI and me very happy.

Reply:
I change `transform_time_offset` for the global EKF for GPS fusion only, but yeah I do use that parameter for this same reason!